File: src/test/java/io/veridia/jsonlogic/TruthyTests.java
package io.veridia.jsonlogic;

import io.veridia.jsonlogic.helpers.ToBoolean;
import org.junit.Test;

import java.util.Collections;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class TruthyTests {
  @Test
  public void testTruthyValues() {
    // Zero
    assertFalse(ToBoolean.eval(0));

    // Any non-zero number
    assertTrue(ToBoolean.eval(1.04));
    assertTrue(ToBoolean.eval(-1));

    // Empty array or collection
    assertFalse(ToBoolean.eval(Collections.EMPTY_LIST));
    assertFalse(ToBoolean.eval(new int[0]));

    // Any non-empty array or collection
    assertTrue(ToBoolean.eval(Collections.singleton(1)));
    assertTrue(ToBoolean.eval(new boolean[] {false}));

    // Empty string
    assertFalse(ToBoolean.eval(""));

    // Any non-empty string
    assertTrue(ToBoolean.eval("hello world"));
    assertTrue(ToBoolean.eval("0"));

    // Null
    assertFalse(ToBoolean.eval(null));

    // NaN and Infinity
    assertFalse(ToBoolean.eval(Double.NaN));
    assertFalse(ToBoolean.eval(Float.NaN));
    assertTrue(ToBoolean.eval(Double.POSITIVE_INFINITY));
    assertTrue(ToBoolean.eval(Double.NEGATIVE_INFINITY));
    assertTrue(ToBoolean.eval(Float.POSITIVE_INFINITY));
    assertTrue(ToBoolean.eval(Float.NEGATIVE_INFINITY));
  }
}

File: src/test/scala/io/veridia/jsonlogic/TruthyTests.scala
package io.veridia.jsonlogic

import io.veridia.jsonlogic.helpers.ToBoolean
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

import java.util as ju

class TruthyTests extends AnyFunSuite with Matchers:

  test("truthy values") {
    ToBoolean.eval(java.lang.Integer.valueOf(0)) shouldBe false
    ToBoolean.eval(java.lang.Double.valueOf(1.04)) shouldBe true
    ToBoolean.eval(java.lang.Integer.valueOf(-1)) shouldBe true

    ToBoolean.eval(ju.Collections.emptyList()) shouldBe false
    ToBoolean.eval(new Array[Int](0)) shouldBe false

    ToBoolean.eval(ju.Collections.singleton(java.lang.Integer.valueOf(1))) shouldBe true
    ToBoolean.eval(Array(false)) shouldBe true

    ToBoolean.eval("") shouldBe false
    ToBoolean.eval("hello world") shouldBe true
    ToBoolean.eval("0") shouldBe true

    ToBoolean.eval(null) shouldBe false

    ToBoolean.eval(java.lang.Double.NaN) shouldBe false
    ToBoolean.eval(java.lang.Float.NaN) shouldBe false
    ToBoolean.eval(java.lang.Double.POSITIVE_INFINITY) shouldBe true
    ToBoolean.eval(java.lang.Double.NEGATIVE_INFINITY) shouldBe true
    ToBoolean.eval(java.lang.Float.POSITIVE_INFINITY) shouldBe true
    ToBoolean.eval(java.lang.Float.NEGATIVE_INFINITY) shouldBe true
  }
